cpp/src/parquet/page_index.h

```diff
@@ -0,0 +1,137 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "parquet/platform.h"
+#include "parquet/statistics.h"
+#include "parquet/types.h"
+
+namespace parquet {
+
+/// \brief BoundaryOrder is a proxy around format::BoundaryOrder.
+enum class PARQUET_EXPORT BoundaryOrder { Unordered = 0, Ascending = 1, Descending = 2 };
+
+/// \brief ColumnIndex is a proxy around format::ColumnIndex.
+class PARQUET_EXPORT ColumnIndex {
+ public:
+  /// \brief Create a ColumnIndex from a serialized thrift message.
+  static std::unique_ptr<ColumnIndex> Make(const ColumnDescriptor& descr,
+                                           const void* serialized_index,
+                                           uint32_t index_len,
+                                           const ReaderProperties& properties);
+
+  virtual ~ColumnIndex() = default;
+
+  /// \brief Returns number of pages in this column index.
+  virtual int64_t num_pages() const = 0;
+
+  /// \brief Returns if all values are null in a single page.
+  virtual bool is_null_page(int64_t page_id) const = 0;
+
+  /// \brief Returns whether both min_values and max_values are
+  /// orderd and if so, in which direction.
+  virtual BoundaryOrder boundary_order() const = 0;
+
+  /// \brief Returns if null count is available.
+  virtual bool has_null_counts() const = 0;
+
+  /// \brief Returns null count for a single page.
+  virtual int64_t null_count(int64_t page_id) const = 0;
+
+  /// \brief The minimum value of a single page. Throws if it is null page.
+  virtual const std::string& encoded_min(int64_t page_id) const = 0;
+
+  /// \brief The maximum value of a single page. Throws if it is null page.
+  virtual const std::string& encoded_max(int64_t page_id) const = 0;
+
+  /// \brief Returns all null indicator for each page in batch.
+  virtual const std::vector<bool>& null_pages() const = 0;
+
+  /// \brief Returns null count for each page in batch.
+  virtual const std::vector<int64_t>& null_counts() const = 0;
+};
+
+/// \brief Typed implementation of ColumnIndex.
+template <typename DType>
+class PARQUET_EXPORT TypedColumnIndex : public ColumnIndex {
+ public:
+  using T = typename DType::c_type;
+
+  /// \brief The minimum value of a single page. Throws if it is null page.
+  virtual T min_value(int64_t page_id) const = 0;
```

Comment:
generally same questions above on API:
1.  Should we return optional for values that might not be available?
2. should we be exposing both Vectors and individual element accessors?

Reply:
1. I think returning optional makes sense and it is better than throwing exception.
2. This is a public-facing API, so I'd keep two kinds of accessors for better usability.